File: Maximum Erasure Value/two-pointers.cpp
#include <iostream>
#include <vector>
#include <numeric>
#include <unordered_set>
using namespace std;

int solution(vector<int>& nums) {
    int n = nums.size();
    vector<int> psum(n+1, 0); 
    partial_sum(nums.begin(), nums.end(), psum.begin()+1);
    
    unordered_set<int> queue;
    int i = 0, score = 0; 
    for(int j=0; j<n; j++) {
        while (queue.find(nums[j]) != queue.end()) {
            queue.erase(nums[i++]);
        }
        score = max(score, psum[j+1] - psum[i]);
        queue.insert(nums[j]);
    }
    return score;
}

int main() {

    // vector<int> nums = {4,2,4,5,6};
    // cout << solution(nums) << endl;

    vector<int> nums = {5,2,1,2,5,2,1,2,5};
    cout << solution(nums) << endl;
}

File: Maximum Erasure Value/two-pointers.py
from itertools import accumulate

def maximumUniqueSubarray(nums):

    psum = [0] + list(accumulate(nums))
    queue = set()
    score = i = 0
    n = len(nums)
    for j in range(n):
        while nums[j] in queue:
            queue.remove(nums[i])
            i += 1

        score = max(score, psum[j+1] - psum[i])
        queue.add(nums[j])

    return score

if __name__ == "__main__":
    
    nums = [4,2,4,5,6]
    maximumUniqueSubarray(nums)

    nums = [5,2,1,2,5,2,1,2,5]
    maximumUniqueSubarray(nums)
